pystachio/choice.py

```diff
@@ -0,0 +1,166 @@
+# Choice types: types that can take one of a group of selected types.
+
+import json
+
+from .base import Object
+from .basic import SimpleObject
+from .compatibility import Compatibility
+from .typing import (
+    Type,
+    TypeCheck,
+    TypeFactory,
+    TypeMetaclass
+)
+
+
+class ChoiceFactory(TypeFactory):
+    """A Pystachio type representing a value which can be one of several
+    different types.
+    For example, a field which could be either an integer, or an integer
+    expression (where IntegerExpression is a struct type) could be written
+    Choice("IntOrExpr", (Integer, IntegerExpression))
+    """
+    PROVIDES = 'Choice'
+
+    @staticmethod
+    def create(type_dict, *type_parameters):
+        """
+        type_parameters should be:
+            (name, (alternative1, alternative2, ...))
+        where name is a string, and the alternatives are all valid serialized
+        types.
+        """
+        assert len(type_parameters) == 2
+        name = type_parameters[0]
+        alternatives = type_parameters[1]
+        assert isinstance(name, Compatibility.stringy)
+        assert isinstance(alternatives, (list, tuple))
+        choice_types = []
+        for c in alternatives:
+            choice_types.append(TypeFactory.new(type_dict, *c))
+        return TypeMetaclass(str(name), (ChoiceContainer,), {'CHOICES': choice_types})
+
+
+class ChoiceContainer(Object, Type):
+    """The inner implementation of a choice type value.
+
+    This just stores a value, and then tries to coerce it into one of the alternatives when
+    it's checked or interpolated.
+    """
+    __slots__ = ('_value',)
+
+    def __init__(self, val):
+        self._value = val
+        self._scopes = ()
+
+    def scopes(self):
+        return self._scopes
+
+    def get(self):
+        if isinstance(self._value, Object):
+            return self._value.get()
+        else:
+            return self._value
+
+    def value(self):
```

Comment:
It looks like it returns an immutable form of the primitive value for some definition of primitive, ie ints and floats and bools for Integers and Floats and Booleans, tuples for Lists, and frozendicts for Maps and Structs.  Agreed that's not  a great value to be testing since it differs from the known `Type`, say `List(String)`.  Ideally I'd think you'd want to be able to test the underlying type directly instead of the value.  Say you had `Choice((List(Integer), List(String)))`) - testing `.value()` requires iterating to the 1st value and concting your own type-test, say `isinstance(l[0], int)`.  It would be more direct to instead say `isinstance(struct.choice_field(), List(Integer))`.  Right now this won't work since `field()` is a `Choice_StringList_IntegerList`, but adding `chosen_type` method as opposed to (or in addtion to) `value` would allow this sort of direct test.